lib/pure/concurrency/cpuinfo.nim

```diff
@@ -101,3 +102,207 @@ proc countProcessors*(): int {.rtl, extern: "ncpi$1".} =
   else:
     result = sysconf(SC_NPROCESSORS_ONLN)
   if result <= 0: result = 0
+
+# TODO(awr1): get cache line size
+# TODO(awr1): get processor name
+# TODO(awr1): ARM processor exts
+# TODO(awr1): on ARM, probe /proc/cpuinfo, but what about windows on arm?
+
+const
+  onX86 = defined(i386) or defined(amd64)
+  onArm = defined(arm) or defined(arm64)
+
+when onX86:
+  proc cpuidX86(leaf: int32): tuple[eax, ebx, ecx, edx: int32] =
+    when defined(vcc):
+      # limited inline asm support in vcc, so intrinsics, here we go:
+      proc cpuidVcc(cpuInfo: ptr int32; functionID: int32)
+        {.cdecl, importc: "__cpuid", header: "intrin.h".}
+      cpuidVcc(addr result.eax, leaf)
+    else:
+      var (eaxr, ebxr, ecxr, edxr) = (0'i32, 0'i32, 0'i32, 0'i32)
+      # zero ecx first!
+      asm """
+        xorl %%ecx, %%ecx
+        cpuid
+        :"=a"(`eaxr`), "=b"(`ebxr`), "=c"(`ecxr`), "=d"(`edxr`)
+        :"a"(`leaf`) """
+      (eaxr, ebxr, ecxr, edxr)
+
+when onX86 or defined(nimdoc):
+  type
+    X86Feature* {.pure.} = enum
+      HypervisorPresence, SimultaneousMultithreading, Vmx, X87fpu, Mmx, MmxExt,
+      F3DNow, F3DNowExt, F3DNowPf, Sse, Sse2, Sse3, Ssse3, Sse4a, Sse41, Sse42,
+      Avx, Avx2, Avx512f, Avx512dq, Avx512ifma, Avx512pf, Avx512er, Avx512cd,
+      Avx512bw, Avx512vl, Avx512vbmi, Avx512vbmi2, Avx512vpopcntdq, Avx512vnni,
+      Avx512vnniw4, Avx512fmaps4, Avx512bitalg, Rdrand, Rdseed, MovBigEndian,
+      Popcnt, Fma3, Fma4, Cas8B, Cas16B, Abm, Bmi1, Bmi2, TsxHle, TsxRtm, Adx,
+      Sgx, Gfni, Aes, Vaes, Vpclmulqdq, Pclmulqdq, NxBit, Float16c, Ssbd,
+      SpecCtrl, Stibp, Sha, Clflush, ClflushOpt, Clwriteback, PrefetchWT1, Mpx
+
+when onArm or defined(nimdoc):
+  type
+    ArmFeature* {.pure.} = enum
+      Neon, Vfp
+
+type CpuFeature* = (when onX86: X86Feature elif onArm: ArmFeature)
+
+proc checkCpuFeatures*(features: set[CpuFeature]): set[CpuFeature] =
+  when onX86:
+    let
+      leaf1 = cpuidX86(leaf = 1)
+      leaf7 = cpuidX86(leaf = 7)
+      leaf8 = cpuidX86(leaf = 0x80000001'i32)
+    for feature in features:
```

Comment:
This should be populated and cached on `import cpuinfo` so that it's done only once for the lifetime of the program.

Applications and libraries that benefits from CPU feature detection target high-performance and repeatedly looping for something that is invariant at runtime is performance left on the table.

Reply:
Generally I was expecting more that the application would hang onto this result (and possibly manipulate it) instead of bringing the set everywhere, but you're right, I'll clean this up